Convert task selection to quick search modal in pipeline builder

karthikjeeyar opened the pull request:
**Fixes**: 
<!-- For e.g Fixes: https://issues.redhat.com/browse/ODC-XXX -->
https://issues.redhat.com/browse/ODC-6133

**Analysis / Root cause**: 
<!-- Briefly describe analysis of US/Task or root cause of Defect -->

Convert the existing task selection dropdown to use the new quick search modal

**Solution Description**: 
<!-- Describe your code changes in detail and explain the solution -->
1. Created a custom PipelineQuicksearchModal. 
2. Extended the quicksearch component to have some configurations like `icon`, `modalPositionOffset` and `isLimitedList`
3. Added autoscroll when cycling through a big list of search results by pressing `down` arrow in the keyboard.
4. Added temporary `Delete` decoarator (This will be changed in subsequent PR's based on the UX design).

Note: This PR contains only the `Task` and `ClusterTask` resources, tekton hub will be done in a separate PR.

**Screen shots / Gifs for design review**: 
<!-- If change affects UI in any way, tag @openshift/team-devconsole-ux and add screenshots/gifs  -->

https://user-images.githubusercontent.com/9964343/126614843-486dd867-fe6d-4a17-920c-03486506f17d.mp4

**test coverage report**: 
<!-- Attach test coverage report -->
Updated task selection method in integration test in pipeline builder page `pipelineBuilder-page.ts` 

**Test setup:**
<!-- If any setup required to test this PR, mention the details -->
Visit Pipeline builder page and click on Add Task node.

**Browser conformance**: 
<!-- To mark tested browsers, use [x] -->
- [x] Chrome
- [x] Firefox
- [ ] Safari
- [ ] Edge

cc: @bgliwa01 @christianvogt @andrewballantyne

karthikjeeyar commented:
cc: @makambalaji Updated the pipeline e2e tests to use quick search modal instead of select dropdown.

makambalaji commented:
> cc: @makambalaji Updated the pipeline e2e tests to use quick search modal instead of select dropdown.

Thanks @karthikjeeyar for the information

andrewballantyne commented on frontend/packages/pipelines-plugin/locales/en/pipelines-plugin.json:
```diff
@@ -3,6 +3,9 @@
   "Create a Tekton Pipeline to automate delivery of your Application": "Create a Tekton Pipeline to automate delivery of your Application",
   "Tasks": "Tasks",
   "Triggers": "Triggers",
+  "Tekton": "Tekton",
```
Translations have a lot of `Tekton` references. Not sure we want this (per my previous comment)

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/catalog/providers/useTasksProvider.tsx:
```diff
@@ -0,0 +1,91 @@
+import { TaskKind } from '../../../types';
+import { TektonTaskAnnotation, TektonTaskLabel } from '../../pipelines/const';
+
+const normalizeTektonTasks = (TektonTasks: TaskKind[], t: TFunction): CatalogItem<TaskKind>[] => {
```
```suggestion
const normalizeTektonTasks = (tektonTasks: TaskKind[]): CatalogItem<TaskKind>[] => {
```

It's a variable, not a class/component name, camelCase please.

Also don't pass `t` into methods.

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/catalog/providers/useTasksProvider.tsx:
```diff
@@ -0,0 +1,91 @@
+    TektonTasks,
+    (acc, task) => {
+      const { uid, name, annotations = {}, creationTimestamp, labels = {} } = task.metadata;
+      const { description } = task?.spec;
```
You just opened yourself up to a NPE on the line above if you're worried about task not being defined :) 

```suggestion
      const { description } = task.spec;
```

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/catalog/providers/useTasksProvider.tsx:
```diff
@@ -0,0 +1,91 @@
+          class: iconClass,
+        },
+        cta: {
+          label: t('pipelines-plugin~Add'),
```
```suggestion
          label: i18next.t('pipelines-plugin~Add'),
```

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/catalog/providers/useTasksProvider.tsx:
```diff
@@ -0,0 +1,91 @@
+
+  const normalizedTektonTasks = React.useMemo(() => normalizeTektonTasks(tektonTasks, t), [
+    tektonTasks,
+    t,
```
So we may want to keep `t` here and eslint ignore. @rohitkrai03 did something similar for his extensions work around kebab items. That way when the lang changes we re-memo them, but we don't make our method reliant on `t` being provided.

It may feel a little odd, but essentially you're splitting "calling functionality" from "memo functionality". We want to recompute on translations for our memo, but we don't want to add the onus on the caller to provide the i18n translator.

divyanshiGupta commented on frontend/packages/console-shared/src/components/quick-search/QuickSearchList.tsx:
```diff
@@ -115,18 +123,21 @@ const QuickSearchList: React.FC<QuickSearchListProps> = ({
-        ))}
-      </div>
+
+      {viewAll.length > 0 && (
```
This will throw error in case viewAll is null/undefined. 
```suggestion
      {viewAll?.length > 0 && (
```

divyanshiGupta commented on frontend/packages/console-shared/src/components/quick-search/QuickSearchList.tsx:
```diff
@@ -115,18 +123,21 @@ const QuickSearchList: React.FC<QuickSearchListProps> = ({
+
+      {viewAll.length > 0 && (
+        <div className="odc-quick-search-list__all-items-link">
+          {viewAll?.map((catalogLink) => (
```
```suggestion
          {viewAll.map((catalogLink) => (
```

divyanshiGupta replied:
Since you are adding the check above we dont need to use optional chaining here.

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-builder/PipelineBuilderForm.tsx:
```diff
@@ -61,8 +64,16 @@ const PipelineBuilderForm: React.FC<PipelineBuilderFormProps> = (props) => {
   useExplicitPipelineTaskTouch();
 
   const statusRef = React.useRef(status);
+  const [container, setContainerView] = React.useState<HTMLElement>(null);
```
Pretty sure this is the wrong hook. Refs are not variables, they are functions that return variables... this is the wrong hook I think.

```suggestion
  const [container, setContainerView] = React.useRef<HTMLElement>(null);
```

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-builder/PipelineBuilderForm.tsx:
```diff
@@ -127,6 +139,14 @@ const PipelineBuilderForm: React.FC<PipelineBuilderFormProps> = (props) => {
             <FormBody flexLayout disablePaneBody className="odc-pipeline-builder-form__grid">
+              <div ref={setContainerView} />
+              <PipelineQuickSearch
+                namespace={'rh-test'}
```
Hmm?

karthikjeeyar replied:
fixed it.

divyanshiGupta commented on frontend/packages/pipelines-plugin/src/components/catalog/providers/useTasksProvider.tsx:
```diff
@@ -0,0 +1,91 @@
+  const normalizedTektonTasks: CatalogItem<TaskKind>[] = _.reduce(
+    TektonTasks,
+    (acc, task) => {
+      const { uid, name, annotations = {}, creationTimestamp, labels = {} } = task.metadata;
```
This will throw error if task is undefined/null
```suggestion
      const { uid, name, annotations = {}, creationTimestamp, labels = {} } = task?.metadata ?? {};
```

andrewballantyne replied:
Not sure metadata can be undefined. And I do not believe we can have null tasks because of the origin is from a request of objects.

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/RemoveNodeDecorator.tsx:
```diff
@@ -0,0 +1,43 @@
+    >
+      <circle cx={0} cy={0} r={BUILDER_NODE_ERROR_RADIUS} fill={greyColor.value} />
+      <g transform="translate(-5, -7)">
+        <foreignObject width={BUILDER_NODE_ERROR_RADIUS * 2} height={BUILDER_NODE_ERROR_RADIUS * 2}>
```
```suggestion
        <foreignObject width={BUILDER_NODE_DECORATOR_RADIUS * 2} height={BUILDER_NODE_DECORATOR_RADIUS * 2}>
```
Maybe we rename the "ERROR" to "DECORATOR" so we can use it for two use-cases without it feeling like we are dipping into some random disassociated number.

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/TaskList.tsx:
```diff
@@ -74,7 +69,9 @@ const TaskList: React.FC<any> = ({
         className="odc-task-list-node__trigger"
-        onClick={() => setMenuOpen(!isMenuOpen)}
+        onClick={() => {
+          setTimeout(() => onTaskSearch(onNewTask), 0); // wait for thread end
```
Hmmm, why do we need to worry about thread end? It's gotta be a shallow thread from an onClick of a 'g' tag.

karthikjeeyar replied:
Right, we don't need to wait on this. fixed the issue.

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/catalog/providers/useTasksProvider.tsx:
```diff
@@ -0,0 +1,91 @@
+}): [CatalogItem[], boolean, string] => {
+  const { t } = useTranslation();
+
+  const { namespacedTasks, clusterTasks } = useK8sWatchResources<{
+    namespacedTasks: TaskKind[];
+    clusterTasks: TaskKind[];
+  }>({
+    namespacedTasks: {
+      kind: referenceForModel(TaskModel),
+      isList: true,
+      namespace,
+    },
+    clusterTasks: {
+      kind: referenceForModel(ClusterTaskModel),
+      isList: true,
+      namespaced: false,
+    },
+  });
```
No access to Formik state from here, eh? Not in the tree so we can't just consume the forms resources?

karthikjeeyar replied:
It is in the tree, I have updated to use it.

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/quicksearch/PipelineQuickSearch.tsx:
```diff
@@ -0,0 +1,93 @@
+  viewContainer?: HTMLElement;
+  isOpen: boolean;
+  setIsOpen: (isOpen: boolean) => void;
+  callback: any;
```
What's the definition of the callback?

karthikjeeyar commented on frontend/packages/pipelines-plugin/console-extensions.json:
```diff
@@ -155,5 +155,27 @@
+    "properties": {
+      "type": "Tekton",
+      "title": "%pipelines-plugin~Tekton%",
+      "catalogDescription": "%pipelines-plugin~Browse for tekton tasks available in the cluster.%"
```
changed catalog-item type to `Red Hat` based on the HLD, we have `Red Hat` and `Community` (Tekton hub).

karthikjeeyar commented on frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/TaskList.tsx:
```diff
@@ -42,17 +37,17 @@ const TaskList: React.FC<any> = ({
     (o) => o.label,
   );
-  const unselectedTaskText = unselectedText || t('pipelines-plugin~Select task');
+  const unselectedTaskText = unselectedText || t('pipelines-plugin~Add task');
```
Updated

karthikjeeyar commented on frontend/packages/console-shared/src/components/quick-search/QuickSearchContent.tsx:
```diff
@@ -32,6 +33,7 @@ const QuickSearchContent: React.FC<QuickSearchContentProps> = ({
   selectedItemId,
   onSelect,
   closeModal,
+  isLimitedList = true,
```
Changed the prop name and made it required.

karthikjeeyar commented on frontend/packages/console-shared/src/components/quick-search/QuickSearchDetails.scss:
```diff
@@ -9,7 +9,7 @@
 
   &__form-button {
-    margin: var(--pf-global--spacer--md) 0px;
+    margin: var(--pf-global--spacer--md) 0px !important;
```
After moving the quick search to console-shared, the margin and highlight background colors are overridden, thus added !important.

divyanshiGupta commented on frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/RemoveNodeDecorator.scss:
```diff
@@ -0,0 +1,4 @@
+.opp-remove-node-decorator {
+  font-size: 10px;
```
```suggestion
  font-size:  var(--pf-global--icon--FontSize--sm);
```

divyanshiGupta reviewed (changes requested):
Code looks good. Will try it locally and add lgtm if everything works fine.

divyanshiGupta commented:
<img width="1791" alt="Screenshot 2021-07-30 at 10 22 03 PM" src="https://user-images.githubusercontent.com/20724543/127686796-a6be993d-824f-4412-8031-10f281188118.png">

@karthikjeeyar In your screen-recording I see the modal is in the centre but if you look at above screenshot it appears on top for me. In topology page the modal appears in the centre so for consistency I think we should keep it in centre in pipeline-builder page as well.

Also in topology quick-search modal we show the provided by label like this `Red Hat, Inc` i.e the first letter is in uppercase. We might want to do the same here.

I also noticed that when there are only five items to show as you can see in the above screenshot there is a weird gap shown after the last item. It's not a big issue but it looks weird. If this can be fixed with minimal effort we can do it in this PR.

divyanshiGupta commented:
https://user-images.githubusercontent.com/20724543/127688057-378d4493-5320-4e71-8674-2d965df6fb1d.mov

@karthikjeeyar I think we want to disable keyboard action for opening the modal as here we dont know where the user might want to add the task i.e they can add it parallelly or side-by-side and if they use the keyboard action to open the modal and add the task it gets added parallelly automatically which is something they might not want.

divyanshiGupta reviewed (changes requested):
@karthikjeeyar I have added a few comments above. Apart from these the functionality works as expected.

karthikjeeyar commented:
> @karthikjeeyar In your screen-recording I see the modal is in the centre but if you look at above screenshot it appears on top for me. In topology page the modal appears in the centre so for consistency I think we should keep it in centre in pipeline-builder page as well.

Initially I have added modalPositionOffset to be at the center of the screen, removed it based on Andrew's [comment here](https://github.com/openshift/console/pull/9583#discussion_r675627922).


> Also in topology quick-search modal we show the provided by label like this Red Hat, Inc i.e the first letter is in uppercase. We might want to do the same here

In the designs we don't have this `Provided by <providername>` at all, it is showing because of the default topology quick search implementation, we have a separate [story](https://issues.redhat.com/browse/ODC-6135) to match the modal content exactly with the final design.

karthikjeeyar commented:
@divyanshiGupta Added a prop to disable keyboard open and used it in pipelineQuicksearch modal

divyanshiGupta commented:
> <img alt="Screenshot 2021-07-30 at 10 22 03 PM" width="1791" src="https://user-images.githubusercontent.com/20724543/127686796-a6be993d-824f-4412-8031-10f281188118.png">
> 
> @karthikjeeyar In your screen-recording I see the modal is in the centre but if you look at above screenshot it appears on top for me. In topology page the modal appears in the centre so for consistency I think we should keep it in centre in pipeline-builder page as well.
> 
> Also in topology quick-search modal we show the provided by label like this `Red Hat, Inc` i.e the first letter is in uppercase. We might want to do the same here.
> 
> I also noticed that when there are only five items to show as you can see in the above screenshot there is a weird gap shown after the last item. It's not a big issue but it looks weird. If this can be fixed with minimal effort we can do it in this PR.

@bgliwa01 wdyt about this?

karthikjeeyar commented:
> I also noticed that when there are only five items to show as you can see in the above screenshot there is a weird gap shown after the last item. It's not a big issue but it looks weird. If this can be fixed with minimal effort we can do it in this PR.

@divyanshiGupta Yes, there isn't a extra space needed here, when we are not showing the viewLinks at the bottom.
Updated the code
![image](https://user-images.githubusercontent.com/9964343/127857216-d5223d08-0edb-46e8-941c-c69bd3487146.png)

divyanshiGupta commented:
@karthikjeeyar I am getting i18n warnings for both TopologyQuickSearch and PipelineQuickSearch.

karthikjeeyar commented:
>  @karthikjeeyar  I am getting i18n warnings for both TopologyQuickSearch and PipelineQuickSearch.
I am not seeing any i18n errors after building the code and re-bridging it. Frontend tests are also passing, so I request you to try build.sh once.

divyanshiGupta commented:
> > @karthikjeeyar In your screen-recording I see the modal is in the centre but if you look at above screenshot it appears on top for me. In topology page the modal appears in the centre so for consistency I think we should keep it in centre in pipeline-builder page as well.
> 
> Initially I have added modalPositionOffset to be at the center of the screen, removed it based on Andrew's [comment here](https://github.com/openshift/console/pull/9583#discussion_r675627922).
> 
> > Also in topology quick-search modal we show the provided by label like this Red Hat, Inc i.e the first letter is in uppercase. We might want to do the same here
> 
> In the designs we don't have this `Provided by <providername>` at all, it is showing because of the default topology quick search implementation, we have a separate [story](https://issues.redhat.com/browse/ODC-6135) to match the modal content exactly with the final design.

@karthikjeeyar I am adding lgtm to this PR since you mentioned we have another story for aligning with UX and the functionality works as expected. But it will be good if you can mention these points in the story so that we can follow up with UX and decide how do we handle these.

/lgtm

divyanshiGupta commented:
/qe-approved
/px-approved
/doc-approved

divyanshiGupta commented:
/label qe-approved
/label px-approved
/label doc-approved

divyanshiGupta commented:
/label docs-approved

invincibleJai commented on frontend/packages/console-shared/src/components/quick-search/QuickSearchList.tsx:
```diff
@@ -57,6 +57,14 @@ const QuickSearchList: React.FC<QuickSearchListProps> = ({
+    if (selectedItemId) {
+      const element = document.getElementById(selectedItemId);
+      if (element) {
+        element?.scrollIntoView({ block: 'nearest' });
```
nit: we can avoid `?` or if condition?

invincibleJai commented:
/lgtm
/approve

verified the changes, works as expected

andrewballantyne commented:
/retest

bgliwa01 commented:
The modal should work like the one in topology so that there is no dark background and it is centered on the screen that is not including the left navigation.

andrewballantyne commented:
> The modal should work like the one in topology so that there is no dark background and it is centered on the screen that is not including the left navigation.

@karthikjeeyar please make the desired changes in your next PR.
